### DISCUSSION TITLE
Implement Oxc transformer from the bottom up.

### DISCUSSION THREAD

**Discussion by Boshen** [oxc-project/oxc#2704] [General]
The current top-to-bottom porting strategy is not working right now. I get depressed whenever I want to implement a plugin due to so many obstacles.

We are going to try another strategy, by doing research and layout all the dependencies of the plugins. After all the dependencies of each plugins are collected, we are going to implement all the utilities from the bottom-up.

Dependencies are anything that we haven't implemented yet.

**Comment by Boshen** [comment]
babel-plugin-transform-typescript:

**Reply by Dunqing** [reply]
These are the tests we still need to solve

https://github.com/oxc-project/oxc/blob/0faa410781261724032914f91f20c46ccf1ea11e/tasks/transform_conformance/babel.snap.md#L835-L896

### Progress (97/158)

- [x] eliminate type or unused import/export
- [x] enum
- [ ] const enum
- [x] namespace
- [ ] nested namespace
- [ ] class constructor
- [ ] decorators (Depends on babel-plugin-proposal-decorators)

### Dependency plugins

- babel-plugin-transform-react-jsx
- babel-plugin-proposal-decorators
- babel-plugin-transform-classes
- babel-plugin-transform-private-methods
- babel-plugin-transform-class-properties
- babel-plugin-transform-object-rest-spread

In addition to this, access to scopes and access to parent nodes are also relied upon in some complex cases.

**Comment by rzvxa** [comment]
### List of APIs needed for implementing transformers(WIP)

- AST
	- Node replacement
		- Replace one
		- Replace one with multiple (replace a node with an array of multiple nodes, especially useful for changing one statement or expression with multiple items of that type)
		- Replace expression with statements (explode statements into hoised variables and expressions)
	- Access to neighbor nodes (can be achieved with #2818)

- Utilities
	- Wrap function (wrap function in a function call, useful for scoping, reference injection. building anonymous functions, etc. 
	- Add helper (lazy loading for helpers, mimicking the babelHelpers generation. We might want to include the whole [babel-helpers](https://github.com/babel/babel/tree/f0a63dbdf64743c00e704277248be1682502e302/packages/babel-helpers/src/helpers) since these are tricky and babel is one of if not the most battle-tested library for transforming js, so they should have these pretty much dialed in)
	- Async to generator
	-  [JSBI](https://github.com/GoogleChromeLabs/jsbi)

- Scope
	- Variable Rename utilities(with random name support to prevent shadowing/conflict)
	- Hoist priority/ordering(might not be needed)
	- Function naming utility(infers the name of a function or class from AST)

- Codegen
	- Configurable generator / codegen plugins(babel has this but it might prove unnecessary)

Since I want to list high-level concepts, I didn't link to their corresponding implementations in Babel or ESbuild. We might want to base them on either one of them or even write something unique to oxc, The thing is that we just need a way to do these operations.

Feel free to edit this comment and help make it more organized and complete.
I am looking forward to your ideas for approaching these!

**Reply by rzvxa** [reply]
I think the replacement part can be achieved by changing how VisitMut works, Or at least how the transformer visitor is working.
Each visit function can at the end return a `Vec<ReplacementNode>` which can either return the same node without modification or return one or many new nodes. This way our transformer's visitors and each transformer act as map functions on our nodes, This Might change a bit while implementing it but it should be a relatively ok pattern for mutating nodes.
When we replace one node with multiple nodes it should be propagated till it hits visiting either a list of expressions or a list of statements which then can be used to replace the singular expr/stmt with multiple nodes. Some places might need to replace a single item with multiple nodes by wrapping them in functions etc.

API functions from Babel can be implemented in this approach as the constructors of `ReplacementNode`(name will change just something to reference it here) for example in theory if we want to replace the current node with 2 expression nodes we should return something like this:

```rust
ReplacementNode::with_expressions(vec![expr1, expr2])
```

The `ReplacementNode` can be either an expression or a statement, It can have the Into trait implemented so it can do the statement explosion mentioned above there, Or maybe we do it the other way around but it can defer this act until we propagate it to a visitor which can append it.

This idea is still incomplete and I'm actively thinking about it, Might have some edge cases that would render it useless. Or I might change it drastically if I get the chance to actually prototype it.

**Reply by Boshen** [reply]
I'm 👍 for making everything more Rusty, but I can't wrap my head around replacing AST nodes while changing scopes and symbols.

I think @milesj mentioned babel path https://github.com/babel/babel/tree/main/packages/babel-traverse/src/path

Do we need to replicate it using a more Rusty approach, otherwise we'll end up with swc's complexity - if we keep the APIs the same then it'll be easier to do a line by line port.

**Reply by milesj** [reply]
I had a similar thought for replacing a node, was something like this:

```rust
impl Transformer for X {
    fn transform(&mut self, node: &AstNode, ctx: &mut TransformationContext)  {
       let node = node.clone();
       // mutate it in some way

       ctx.replaceNode(node)
           .removeBinding("name")
           .renameBinding("old", "new")
           .insertBinding("another");
    }
}
```

I think having our transforms methods return `Option<Node>`, or have another API that tracks this, is maybe best for handling the mutations.

**Reply by rzvxa** [reply]
@milesj It is nice to have it in our context since it would make the API similar to the Babel, And won't enforce the return values on all mutable visitors(I wanted to get away with defining an Output type alias in our VisitMut which can be set to unit to make this feature available for those visitors who actually wish to do such things. However, Doing it in context might make some parts more complicated especially if we go for a single AST instead of having 2 copies(one mutable and one immutable). How do we plan to apply these tree manipulations in the ctx? Do we mark ctx as dirty and apply it after the transformation? Or is there some different approach that I'm missing here?

**Reply by rzvxa** [reply]
> Do we need to replicate it using a more Rusty approach, otherwise we'll end up with swc's complexity - if we keep the APIs the same then it'll be easier to do a line by line port.

I strongly believe that with the addition of the `ast_node_id` and a trait to generalize it to all types we might actually get away with it without any real need for the `Path` from Babel, Even the API can stay similar since we can do most things happening in Path via our node_id and semantic data(for example getting the parent node, scope data, variable bindings, symbols, etc). The removal part and mutation, in general, should be done in a different manner since we can't replace nodes without sacrificing safety here. Maybe for mutation, we can use one of the approaches suggested by @milesj?

**Reply by milesj** [reply]
> Do we mark ctx as dirty and apply it after the transformation? Or is there some different approach that I'm missing here?

The way I was thinking about this is that the context argument is a new instance for every node, which only tracks state for that node. It could then tell if it's dirty or not, and then apply whatever changes back to the service. Here's some example pseudo-code.

```rust
impl<'a> VisitMut<'a> for X<'a> {
	fn visit_await_expression(&mut self, expr: &mut AwaitExpression<'a>) {
		let mut ctx = TransformationContext::new();

		self.run_enter_transforms(expr, &ctx);

		if ctx.is_dirty() {
			self.apply_changes(&ctx);
			*expr = ctx.replaced_node();
		}

		walk_await_expression_mut(self, expr);

		let mut ctx = TransformationContext::new();

		self.run_leave_transforms(expr, &ctx);

		if ctx.is_dirty() {
			self.apply_changes(&ctx);
			*expr = ctx.replaced_node();
		}

		if ctx.should_retraverse() {
			self.visit_await_expression(expr);
		}
	}
}
```

Not sure if this is actually viable though. It would also require a bunch of boiler-plate for every `VisitMut` method.

**Reply by rzvxa** [reply]
> The way I was thinking about this is that the context argument is a new instance for every node, which only tracks state for that node. It could then tell if it's dirty or not, and then apply whatever changes back to the service. Here's some example pseudo-code.

I find it really clean and easy to work with, We might be able to get around the boilerplate using some declarative macros, For example doing something like this:

```rust
impl<'a> VisitMut<'a> for X<'a> {
	fn visit_await_expression(&mut self, expr: &mut AwaitExpression<'a>) {
		transform!(self.run_enter_transforms(expr));

		walk_await_expression_mut(self, expr);

		let ctx = transform!(self.run_leave_transforms(expr));

		if ctx.should_retraverse() {
			self.visit_await_expression(expr);
		}
	}
}
```

I was going to do something similar using macros that would act on the return value, But instead, we can have a macro that defines a context, passes it in as the first argument, and applies the result, At the end it would return the ctx so we can act on it if we need some information from it.

I'm just not sure how we can use this pattern to replace a parent node, Do we have to use it with a combination of other tricks?

> No child -> parent traversal, instead parent -> child visit pre-checks

I think it can work great with this ideology which I find intriguing if we find a workaround for performance loss(we need some extra checks that can be mitigated with unreachable default conditions).
If we had a VisitMutContext it would've changed everything since we could actually propagate the parent change to our upstream visit method.

**Reply by milesj** [reply]
> I'm just not sure how we can use this pattern to replace a parent node, Do we have to use it with a combination of other tricks?

You wouldn't. You would need to do the 2nd or 3rd proposal in my post (I like the 2nd one). But basically you would mutate both the parent and child separately, but set state between the 2 to determine when and what to do.

**Reply by rzvxa** [reply]
Well, I think that abstracting changes to be done using the ctx can be a good thing, There is not much difference in their complexity when it comes to the implementation but the API with ctx can be more extensive and do a lot more if needed.

For mutation, I like the third option a bit more than the second one and I'm not that fond of the first approach. The second approach is similar to the third with extra steps, Maybe I'm not getting it completely?

**Reply by milesj** [reply]
@rzvxa Yeah 2 and 3 are very similar, but the only difference I'd say is this.

For 3, if your parent has a deep child tree, then you're doing a ton of traversal to figure out if a certain child exists within that tree. However, there's no re-traversal after mutating the parent node.

For 2, it's the opposite. Since you're already in the child, you just need to loop through the handful of parent nodes to see if it's within a certain parent. However, if it is in a parent, then once we go back up the tree to the parent node, then we need to re-traverse down again if we mutated it.

Honestly, we can probably support both patterns since they are so similar, and depending on how complex a rule is, they can choose either approach.

**Reply by milesj** [reply]
I'll start spiking this if you want to focus on the ast node IDs.

**Reply by rzvxa** [reply]
Yes, It would be great if we start working in parallel on these 2 subjects. I suspect that the ast node ID won't take that long since it's mostly grunt work and everything is already figured(at least as far as we know at the moment). So I get done with it and join you on the mutability train.

**Comment by milesj** [comment]
Here's my notes and thoughts on this matter.

------

Requirements

- [ ] Traverse upwards through parents
- [ ] Be able to mutate parents via child, or vice versa
- [ ] Scopes/symbols/etc need to be updated when something is mutated

Nice to haves

- [ ] Traverse sideways through siblings (can be done via parent downwards)
- [ ] Parallelization (per file)

# Different patterns for mutating

Some ideas I had while trying to think of a new transformer implementation.

### Immutable AST builds a mutable AST on-demand

With this pattern, we would have 2 ASTs, the 1st is the original immutable AST, and the 2nd is a new mutable AST (starts at `Program`) that gets built on-demand while visiting the original AST. Nodes are cloned from the 1st into the 2nd, and only mutated in the 2nd. Cloning should be cheap since we use lifetimes/references for inner string data.

Pros

- Easy to understand
- Doesn't mutate the original AST

Cons

- Traversing/mutating parent still very difficult
- How to keep semantic in sync???

```rust
impl Transformer for X {
    fn transform(&mut self, old_node: &AstNode) -> Result<AstNode> {
        let mut new_node = old_node.clone();
        // Mutate node

        Ok(new_node)
    }
}
```

### Enter/Leave specific transform methods

Instead of having a single `transform(in) -> out` method, we would have 2 methods, `enter(in) -> out` and `leave(in) -> out`, which would fire when traversing down the tree, and then back upwards. With this pattern, if a child must mutate its parent, it can set a flag in `enter`, and then in `leave`, it can handle the mutation based on the flag.

Pros

- Easy to understand
- Mutates in place
- Traversing/mutating parent is easier

Cons

- If the parent mutates the children in `leave`, it will need to be re-traversed. (babel re-traverses)
- How to keep semantic in sync???

```rust
impl Transformer for X {
    fn transform(&mut self, node: &AstNode) -> Result<Option<AstNode>> {
        match node.kind() {
            SomeNode => {
                if self.ctx.is_within_function(...) {
                    // Flag parent changes within child
                    self.in_function_body = true;
                }
            }
        };

        Ok(None)
    }

    fn transform_on_leave(&mut self, node: &AstNode) -> Result<Option<AstNode>> {
        if self.in_function_body && node.kind() == Function {
            let mut node = node.clone();
            // Mutate parent and return a replacement node

            return Ok(Some(node));
        }

        Ok(None)
    }
}
```

### No child -> parent traversal, instead parent -> child visit pre-checks

> This is how TypeScript (and I believe esbuild) work!

This pattern goes against our requirement of "traversing up to the parent from the child", but instead offers an alternative. If we have situations where a child needs to detect if it's within a certain parent, or that parent needs to be mutated base on the status of a child, we can solve these by implementing a custom `Visit` and running it within the parent.

Pros

- Existing transformers will still work the same
- No reverse traversal necessary
- Mutating the parent _before_ mutating the child, avoids re-traversals

Cons

- Requires traversing the children more than necessary (probably fine)
- More work in the parent instead of the child
- Possibly more overhead
- How to keep semantic in sync???

```rust
struct IsChildInParent {
    pub yes: bool,
}

impl Visit for IsChildInParent {
    fn enter_node(&mut self, kind: &AstKind) {
        match kind {
            Child1 | Child2 => {
                self.yes = true;
            }
            _ => {}
        };
    }
}

impl Transformer for X {
    fn transform(&mut self, node: &AstNode) -> Result<Option<AstNode>> {
        match node.kind() {
            ParentNode => {
                let visitor = IsChildInParent { yes: false };
                visitor.visit_parent_node(&ast.node);

                if visitor.yes {
                    // Child exists within this parent,
                    // so mutate the parent from this context
                }
            }
            ChildNode => {
                // Mutate child if need be
            }
        };

        Ok(None)
    }
}
```

# Open questions

Based on all of this, the big problem I keep coming back to, is how do we keep the semantic tables in sync? Especially when we need to add/remove/rename/move scopes/bindings/symbols/etc.

This is further made complicated by the fact that `Semantic` is immutable, and is built entirely around node/scope IDs, but when we are traversing the AST (`Visit` or `VisitMut`), we actually don't have those IDs.

**Reply by rzvxa** [reply]
> Based on all of this, the big problem I keep coming back to, is how do we keep the semantic tables in sync? Especially when we need to add/remove/rename/move scopes/bindings/symbols/etc.

Do we need to do this though? If we keep all transformers in a single traverse step, We can freely change the semantics as we lower the AST, Without worrying about synchronization or any race conditions.

```rust
    fn visit_expression(&mut self, expr: &mut Expression<'a>) {
        // self.typescript.as_mut().map(|t| t.transform_expression(expr));
        self.react_jsx.as_mut().map(|t| t.transform_expression(expr));
        self.regexp_flags.as_mut().map(|t| t.transform_expression(expr));

        self.es2021_logical_assignment_operators.as_mut().map(|t| t.transform_expression(expr));
        self.es2020_nullish_coalescing_operators.as_mut().map(|t| t.transform_expression(expr));
        self.es2015_arrow_functions.as_mut().map(|t| t.transform_expression(expr));
        self.es2015_instanceof.as_mut().map(|t| t.transform_expression(expr));
        self.es2016_exponentiation_operator.as_mut().map(|t| t.transform_expression(expr));
        self.es2015_template_literals.as_mut().map(|t| t.transform_expression(expr));
        self.es2015_new_target.as_mut().map(|t| t.transform_expression(expr));

        walk_expression_mut(self, expr);
    }

```

Given the visit method above, Each transformer would change the semantics as needed and return to the visitor for the next transformation on that node(if not already replaced). Since we have the semantic as `Rc<RefCell<Semantics<'a>>>` in our context It should provide mutation without any extra steps.
The actual semantic generation can be split into SemanticBuilder and SemanticHelpers which the ladder would provide underlying implementations used in SemanticBuilder, We can use these helpers to also transform our semantics(things like adding and removing from the scope or renaming it should be done in helpers instead of directly being implemented in the builder.

> This is further made complicated by the fact that Semantic is immutable, and is built entirely around node/scope IDs, but when we are traversing the AST (Visit or VisitMut), we actually don't have those IDs.

Would you give #2818 a read, It would resolve this issue with only one numeric value added to AST node structures(the actual nodes not the AstNode type), AstNode type can be deprecated if we move on with #2818. Keep in mind for Parallelization we would want to change AstNodeId to contain 2 indexes similar to ESBuild, One pointing to the source index(We assume each source runs on a different thread even if it is not) and one to the actual scope/node ID.

**Reply by milesj** [reply]
@rzvxa Honestly, it's a bit hard to say until we start building these out. But for example, off the top of my head, say we have a class and it has a scope ID of 3, and we down-level it to an IIFE which adds additional scopes (just guessing). Do we keep the scope ID of 3, or do we wipe it away and realign the new scopes with new IDs? How does this affect children/parent hierarchy stuff? What happens to symbols/bindings that were in scope 3?

The problem with `Rc<RefCell<Semantics<'a>>>` is that it's a representation of the file _before_ we transform it. So once we mutate scopes/bindings/symbols/etc, or move nodes around, that struct/table is now incorrect. I experimented with using `SemanticBuilder` here (https://github.com/milesj/oxc/pull/1) but ran into some other issues now that `Visit` and `VisitMut` have incompatible APIs.

> Would you give https://github.com/oxc-project/oxc/issues/2818 a read, It would resolve this issue with only one numeric value added to AST node structures(the actual nodes not the AstNode type), AstNode type can be deprecated if we move on with https://github.com/oxc-project/oxc/issues/2818. Keep in mind for Parallelization we would want to change AstNodeId to contain 2 indexes similar to ESBuild, One pointing to the source index(We assume each source runs on a different thread even if it is not) and one to the actual scope/node ID.

+1 to adding the ID to nodes. I think this would solve a handful of problems.

**Reply by rzvxa** [reply]
> +1 to adding the ID to nodes. I think this would solve a handful of problems.

If I'm not mistaken it would even solve our problem with the difference between `Visit` and `VisitMut`, Since we won't need to pass the AstKind or AstNode anymore, we can handle these by passing ID which can be the same in both traits and provides access to both AstNode and as a result the AstKind.

In theory, it should allow us to use SemanticHelpers/SemanticBuilder to rewrite our semantics.

**Reply by milesj** [reply]
Yup exactly. I wonder if we should also include scope ID on applicable nodes...

**Reply by rzvxa** [reply]
If we go for having the `AstNodeId` accessible on the node itself, We can remove the current `AstNode` type, Make `AstNodes` SOA, and add a mapping between AstNodeId and its ScopeId. So without having more information on nodes with only one ID(or two for thread safety), We can access anything related to that node, It being scope data, symbol, or any other things that would get added later on even if not via semantics(let's say a custom semantic data used for a specific framework), We just map the new information to our AstNodeId and access it with a simple array indexing.

**Reply by rzvxa** [reply]
> Honestly, it's a bit hard to say until we start building these out. But for example, off the top of my head, say we have a class and it has a scope ID of 3, and we down-level it to an IIFE which adds additional scopes (just guessing). Do we keep the scope ID of 3, or do we wipe it away and realign the new scopes with new IDs? How does this affect children/parent hierarchy stuff? What happens to symbols/bindings that were in scope 3?

Just to make it clear, I didn't miss this section. Right now similar to you I don't really have a clear answer to it, But I guess the index order wouldn't make any difference in our result, I have to either try to implement such a thing to make sure Or maybe we can find a similar case in babel or esbuild to see what is their approach there. I can't recall any transformation with the need for a new scope within Babel, However, I might be wrong and have to specifically look for it to be sure. Although I've seen moving bindings from one scope to another pretty much all over the place.

**Reply by milesj** [reply]
I'm of the same thought. We may just be over-thinking this and may not need it. Sounds like one of those features we could maybe push off until we run into a situation where we actually have to support it. At minimum, we just need a way for semantic to be mutable.

**Reply by rzvxa** [reply]
> I'm of the same thought. We may just be over-thinking this and may not need it. Sounds like one of those features we could maybe push off until we run into a situation where we actually have to support it. At minimum, we just need a way for semantic to be mutable.

Just to be sure, You are talking about custom semantic data? Or the idea of having ast_node_id in general? If it is about custom semantics, It would be just a fortunate side effect and not the goal.

**Reply by milesj** [reply]
The `Semantic` instance, or however we end up tracking symbols/etc.

**Reply by rzvxa** [reply]
> The Semantic instance, or however we end up tracking symbols/etc.

I agree, We may want to shift our approach as we implement it. We can start simply by only extracting needed functions from SemanticBuilder into a semantic helper utility since it should get things working and if we need better abstraction we can always build on top of it.

**Reply by hyf0** [reply]
I was thinking you might not need to sync `Semantic`. `Semantic` is very useful for linter but might be over engineering for transformer. There's a lot of stuffs of `Semantic` is useless to transformer, but as you said to sync `Semantic` while mutating the AST, it would be even more complex to maintain the correctness of useless parts of `Semantic`.

Thus, I was thinking maybe we could have a low-level `Semantic` API which is just designed for transformer. Maybe `Semantic` could be built on top of it for reusing some works.

However, it's just a thought of my head and don't have something specific.

**Reply by rzvxa** [reply]
@hyf0 You are right about semantics not being that important in the transformation section we for sure can do without most of the semantic information, But as a library, it would be nice to have the updated semantics after the transformation in case some oxc users want to have some operations based on semantic after transforming their code to the target platform.

However, we are currently trying to reinvent the transformers, so we might go either way depending on the current requirements and how long it will take to do it in the "right way".

We most likely are going to separate SemanticBuilder from utilities needed for mutation of the semantics this way we can have manual mutations using these utilities when transforming the code, So we don't need to worry about the SemanticBuilder and we just add or remove a few identifiers, symbols and scopes in the semantics we got.

However, We can also go without updating the semantics and let the end user generate a new semantic if needed after the transformations. Requires more work if semantics is needed since now we have to walk the whole tree to create the new semantics of the transformed result but might simplify the transformations.

All these bring me to my last point, Having semantics helps us validate the transformations since we need to keep track of old variables for not shadowing them, if we do not mutate the semantics will require us to keep track of applied changes in some other manner.
